### PR TITLE
Changes to sessionization logic and price extraction

### PR DESCRIPTION
**Pull request by sarveshnagpal**
- Moving to session cookie instead of session storage since it's unreliable on iOS Safari
- Bumping up the version to 0.6.11
- Adding support for TabId as part of a dimension (still through Session Storage)
- Updating cookie handling to be forward compatible for future changes

### CODE REVIEW

File: packages/clarity-js/types/data.d.ts [hunk 1]
```diff
@@ -92,7 +92,9 @@ export const enum Dimension {
     ProductCategory = 11,
     ProductSku = 12,
     ProductCurrency = 13,
-    ProductCondition = 14
+    ProductCondition = 14,
+    TabId = 15,
```
Comment (toby-walker) [inline packages/clarity-js/types/data.d.ts:96]: Shouldn't this go to the envelope? Doesn't seem like it belongs here.

File: packages/clarity-js/types/data.d.ts [hunk 2]
```diff
@@ -134,7 +136,8 @@ export const enum BooleanFlag {
 
 export const enum Setting {
     Expire = 365, // 1 Year
-    SessionTimeout = 30 * Time.Minute, // 10 minutes
+    SessionExpire = 1, // 1 Day
+    SessionTimeout = 30 * Time.Minute, // 30 minutes
```
Comment (toby-walker) [inline packages/clarity-js/types/data.d.ts:140]: Why shrink?

File: packages/clarity-js/types/core.d.ts [hunk 3]
```diff
@@ -102,6 +102,7 @@ export interface Config {
     regions?: Regions;
     metrics?: Metrics;
     cookies?: string[];
+    server?: string;
```
Comment (toby-walker) [inline packages/clarity-js/types/core.d.ts:105]: What if this is empty/undefined?

File: packages/clarity-js/src/data/upload.ts [hunk 4]
```diff
@@ -126,8 +126,8 @@ function stringify(encoded: EncodedPayload): string {
 
 function send(payload: string, sequence: number, last: boolean): void {
     // Upload data if a valid URL is defined in the config
-    if (typeof config.upload === Constant.String) {
-        const url = config.upload as string;
+    if (typeof config.upload === Constant.String && config.server) {
```
Comment (toby-walker) [inline packages/clarity-js/src/data/upload.ts:129]: Thsi seems like it wouldn't be backward compatible. Will stop uploading everything on older config.

File: packages/clarity-js/src/data/metric.ts [hunk 5]
```diff
@@ -5,7 +5,9 @@ import encode from "./encode";
 export let data: MetricData = null;
 export let updates: MetricData = null;
 let metricMap: WeakMap<Node, string> = null; // Maps metric nodes => innerText
-const regex = /[^0-9\.]/g;
+const numberRegex = /[^0-9\.,]/g;
+const digitsRegex = /[^0-9\.]/g;
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metric.ts:9]: What's the diff between a number regex and a digits regex in semantics? I see the difference in regex, but not clear what you intend from names.

File: packages/clarity-js/src/data/metric.ts [hunk 6]
```diff
@@ -5,7 +5,9 @@ import encode from "./encode";
 export let data: MetricData = null;
 export let updates: MetricData = null;
 let metricMap: WeakMap<Node, string> = null; // Maps metric nodes => innerText
-const regex = /[^0-9\.]/g;
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metric.ts:8]: NAive JS question: Is it more efficient to put the consts here rahter than in the function that use them (where it's easier to read the code as they are next to each other)?

File: packages/clarity-js/src/data/metadata.ts [hunk 7]
```diff
@@ -19,19 +19,29 @@ export function start(): void {
...
-    sessionId: s.id,
+    sessionId: s.session,
     pageNum: s.count
   }
 
+  // For backward compatibility; remove in future iterations (v0.6.11)
+  if (config.upload && typeof config.upload === Constant.String && (config.upload as string).indexOf(Constant.HTTPS) === 0) {
+    let url = config.upload as string;
+    config.server = url.substr(0, url.indexOf("/", Constant.HTTPS.length));
+    config.upload = config.server.length > 0 && config.server.length < url.length ? url.substr(config.server.length + 1) : url;
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metadata.ts:30]: >&& [](start = 45, length = 2)

All this string munging is hard to follow. What are you doing here? Can you add a comment about what you this is for?

Comment (toby-walker) [inline packages/clarity-js/src/data/metadata.ts:30]: Why do you truncate the URL to the server name only in this?

File: packages/clarity-js/src/data/metadata.ts [hunk 8]
```diff
@@ -71,17 +81,27 @@ export function consent(): void {
 }
 
 export function clear(): void {
-  // Clear any stored information in the session so we can restart fresh the next time
-  if (supported(window, Constant.SessionStorage)) { sessionStorage.removeItem(Constant.StorageKey); }
+  // Clear any stored information in the session cookie so we can restart fresh the next time
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metadata.ts:84]: >session [](start = 41, length = 7)

session cookie is always a bit amgiuous between (1) a permanent cookie that hold the session id and (2) a session-scoped (lifetime) cookie. Reword to clarify which you mean.

File: packages/clarity-js/src/data/metadata.ts [hunk 9]
```diff
@@ -71,17 +81,27 @@ export function consent(): void {
...
+  setCookie(Constant.SessionKey, Constant.Empty, 0);
+}
+
+function tab(): string {
+  let id = shortid();
+  if (supported(window, Constant.SessionStorage)) {
+    let value = sessionStorage.getItem(Constant.TabKey);
+    id = value && value.indexOf(Constant.Separator) < 0 ? value : id;
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metadata.ts:92]: >value [](start = 18, length = 5)

What's the separator for if this doesn't use it. Backwards compatible?

**Comment by toby-walker** [comment]
What happens in this scenario:

1) Old versoin from one DC
2) This version from another DC
3) Back to old version from another DC

If the DC rollout hasn't fniished so tha we have the same browser "flapping" between old and new versions?

File: packages/clarity-js/src/data/metadata.ts [hunk 10]
```diff
@@ -90,13 +110,7 @@ function supported(target: Window | Document, api: string): boolean {
...
 
 function track(): void {
-  if (config.track && supported(document, Constant.Cookie)) {
-    let expiry = new Date();
-    expiry.setDate(expiry.getDate() + Setting.Expire);
-    let expires = expiry ? Constant.Expires + expiry.toUTCString() : Constant.Empty;
-    let value = `${data.userId}${Constant.Semicolon}${expires}${Constant.Path}`;
-    document.cookie = Constant.CookieKey + Constant.Equals + value;
-  }
+  setCookie(Constant.CookieKey, data.userId, Setting.Expire);
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metadata.ts:113]: Don't we lose the structure of the cookie this way?

File: packages/clarity-js/src/data/metadata.ts [hunk 11]
```diff
@@ -108,17 +122,16 @@ function shortid(): string {
...
-    let value = sessionStorage.getItem(Constant.StorageKey);
-    if (value && value.indexOf(Constant.Separator) >= 0) {
-      let parts = value.split(Constant.Separator);
-      if (parts.length === 5 && output.ts - num(parts[1]) < Setting.SessionTimeout) {
-        output.id = parts[0];
-        output.count = num(parts[2]) + 1;
-        output.upgrade = num(parts[3]);
-        output.upload = parts[4];
-      }
+  let output: Session = { session: shortid(), ts: Math.round(Date.now()), count: 1, upgrade: BooleanFlag.False, upload: Constant.Empty };
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metadata.ts:125]: What happened to the condition on config.track?

File: packages/clarity-js/src/data/metadata.ts [hunk 12]
```diff
@@ -129,11 +142,12 @@ function num(string: string, base: number = 10): number {
 }
 
 function user(): string {
-  let id = cookie(Constant.CookieKey);
-  return id && id.length > 0 ? id : shortid();
+  let id = getCookie(Constant.CookieKey);
+  // Splitting and looking up first part for forward compatibility, in case we wish to store additional information in a cookie
+  return id && id.length > 0 ? id.split(Constant.Separator)[0] : shortid();
```
Comment (toby-walker) [inline packages/clarity-js/src/data/metadata.ts:147]: >Separator [](start = 49, length = 9)

CookieSeparator to make it clear what is a separator for (shortened like CookieSep or something is fine).

**Review by toby-walker — CHANGES_REQUESTED** [review]
:clock1:

File: packages/clarity-js/types/core.d.ts (hunk 3 above)
Comment (sarveshnagpal) [reply packages/clarity-js/types/core.d.ts:105]: Then that's fine - it will be handled like how we handled empty upload (or upload callback) scenario.

File: packages/clarity-js/types/data.d.ts (hunk 2 above)
Comment (sarveshnagpal) [reply packages/clarity-js/types/data.d.ts:140]: Not expanding... only updating the comment to reflect the true value so comment stays accurate.

File: packages/clarity-js/src/data/metadata.ts [hunk 13]
```diff
@@ -19,19 +19,29 @@ export function start(): void {
...
+    sessionId: s.session,
     pageNum: s.count
   }
 
+  // For backward compatibility; remove in future iterations (v0.6.11)
+  if (config.upload && typeof config.upload === Constant.String && (config.upload as string).indexOf(Constant.HTTPS) === 0) {
+    let url = config.upload as string;
+    config.server = url.substr(0, url.indexOf("/", Constant.HTTPS.length));
+    config.upload = config.server.length > 0 && config.server.length < url.length ? url.substr(config.server.length + 1) : url;
+  }
```
Comment (swaathee) [inline packages/clarity-js/src/data/metadata.ts:31]: any reason why we cant just keep track of clarity-js version in the session storage for the backward forward compatibility - if the version is not-updated you can send to old session DC else send to new one (and reset session?)

File: packages/clarity-js/types/data.d.ts (hunk 1 above)
Comment (sarveshnagpal) [reply packages/clarity-js/types/data.d.ts:96]: Thought about it and decided to not include it in the envelope because it has no bearing on aggregation. It's a hint for visualization - if we choose to use it later.

File: packages/clarity-js/src/data/metadata.ts [hunk 14]
```diff
@@ -108,17 +122,16 @@ function shortid(): string {
...
-        output.count = num(parts[2]) + 1;
-        output.upgrade = num(parts[3]);
-        output.upload = parts[4];
-      }
+  let output: Session = { session: shortid(), ts: Math.round(Date.now()), count: 1, upgrade: BooleanFlag.False, upload: Constant.Empty };
+  let value = getCookie(Constant.SessionKey);
+  if (value) {
+    let parts = value.split(Constant.Separator);
+    if (parts.length === 5 && output.ts - num(parts[1]) < Setting.SessionTimeout) {
+      output.session = parts[0];
```
Comment (swaathee) [inline packages/clarity-js/src/data/metadata.ts:130]: >session [](start = 13, length = 7)

how are we dealing with overlapping pageviews across tabs?

File: packages/clarity-js/src/data/metric.ts (hunk 5 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metric.ts:9]: I will rename the variables to be more clear. In one case we remove commas (for en fallback if Intl.NumberFormat is not present) and in another we keep the comma to handle it properly respecting page language.

File: packages/clarity-js/src/data/metric.ts (hunk 6 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metric.ts:8]: It's a good question. It's possible things have improved over time and browsers are smart about handling it internally now but here's a link that provides some additional info: https://stackoverflow.com/questions/8814009/how-often-does-javascript-recompile-regex-literals-in-functions

File: packages/clarity-js/src/data/metadata.ts (hunk 7 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:30]: Now that we use cookie to store session state - it seemed wasteful to cache the complete URL. Especially since cookies are sent along with every network requests. This is backward compatibility code. And in future, we can simplify by sending server variable once and then report or upload url can be relative paths.

File: packages/clarity-js/src/data/metadata.ts (hunk 8 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:84]: Sounds good. Will reword.

File: packages/clarity-js/src/data/metadata.ts [hunk 15]
```diff
@@ -71,17 +81,27 @@ export function consent(): void {
...
 
 export function save(): void {
   let ts = Math.round(Date.now());
   let upgrade = config.lean ? BooleanFlag.False : BooleanFlag.True;
   let upload = typeof config.upload === Constant.String ? config.upload : Constant.Empty;
   if (upgrade && callback) { callback(data, !config.lean); }
-  if (config.track && supported(window, Constant.SessionStorage)) {
-    sessionStorage.setItem(Constant.StorageKey, [data.sessionId, ts, data.pageNum, upgrade, upload].join(Constant.Separator));
+  if (config.track && supported(document, Constant.Cookie)) {
+    setCookie(Constant.SessionKey, [data.sessionId, ts, data.pageNum, upgrade, upload].join(Constant.Separator), Setting.SessionExpire);
```
Comment (swaathee) [inline packages/clarity-js/src/data/metadata.ts:104]: this check is present in the setcookie function anyway

File: packages/clarity-js/src/data/metadata.ts [hunk 16]
```diff
@@ -71,17 +81,27 @@ export function consent(): void {
...
 export function clear(): void {
-  // Clear any stored information in the session so we can restart fresh the next time
-  if (supported(window, Constant.SessionStorage)) { sessionStorage.removeItem(Constant.StorageKey); }
+  // Clear any stored information in the session cookie so we can restart fresh the next time
+  setCookie(Constant.SessionKey, Constant.Empty, 0);
+}
+
+function tab(): string {
+  let id = shortid();
+  if (supported(window, Constant.SessionStorage)) {
```
Comment (swaathee) [inline packages/clarity-js/src/data/metadata.ts:90]: >supported [](start = 6, length = 9)

should we have config.track check here?

File: packages/clarity-js/src/data/metadata.ts (hunk 10 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:113]: Sorry not sure if I fully follow. I'm keeping the same format for user cookie as before. Did I miss something here?

File: packages/clarity-js/src/data/metadata.ts (hunk 12 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:147]: I will change it to Pipe which is what the character is.

File: packages/clarity-js/src/data/metadata.ts (hunk 14 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:130]: It's handled in the visualization right now.

File: packages/clarity-js/src/data/metadata.ts (hunk 15 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:104]: Good point, will remove, thanks.

File: packages/clarity-js/src/data/metadata.ts (hunk 16 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:90]: I thought tracking tab is not the same as tracking user or session. But, I do see your point. Let me add the check.

File: packages/clarity-js/src/data/metric.ts (hunk 5 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metric.ts:9]: Done.

File: packages/clarity-js/src/data/metadata.ts (hunk 7 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:30]: I'll add comments to explain it more. I am hoping we can remove this bit with the next version.

Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:30]: Done

File: packages/clarity-js/src/data/metadata.ts (hunk 8 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:84]: Done

File: packages/clarity-js/src/data/metadata.ts (hunk 12 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:147]: Done

File: packages/clarity-js/src/data/metadata.ts (hunk 15 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:104]: Done

File: packages/clarity-js/src/data/metadata.ts (hunk 16 above)
Comment (sarveshnagpal) [reply packages/clarity-js/src/data/metadata.ts:90]: Done

**Review by swaathee — APPROVED** [review]
:shipit:

**Comment by sarveshnagpal** [comment]
Synced offline with Swaathee and verified the changes work fine in all permutations (that we could think of). 

---
In reply to: [816971978](https://github.com/microsoft/clarity/pull/127#issuecomment-816971978) [](ancestors = 816971978)

**Review by sarveshnagpal — COMMENTED** [review]
Tested the changes with multiple permutations.
